ibm/resource_ibm_appid_application.go

```diff
@@ -0,0 +1,216 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringLenBetween(1, 50),
+			},
+			"type": {
+				Description:  "The type of application to be registered. Allowed types are `regularwebapp` and `singlepageapp`, default is `regularwebapp`.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Optional:     true,
+				Default:      "regularwebapp",
+				ValidateFunc: validation.StringInSlice([]string{"regularwebapp", "singlepageapp"}, false),
+			},
+			"secret": {
+				Description: "The `secret` is a secret known only to the application and the authorization server",
+				Type:        schema.TypeString,
+				Computed:    true,
+				Sensitive:   true,
+			},
+			"oauth_server_url": {
+				Description: "Base URL for common OAuth endpoints, like `/authorization`, `/token` and `/publickeys`",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"profiles_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"discovery_endpoint": {
+				Description: "This URL returns OAuth Authorization Server Metadata",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func resourceIBMAppIDApplicationCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	appIDClient, err := meta.(ClientSession).AppIDAPI()
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	tenantID := d.Get("tenant_id").(string)
+	appName := d.Get("name").(string)
+	appType := d.Get("type").(string)
+
+	input := &appid.RegisterApplicationOptions{
+		TenantID: &tenantID,
+		Name:     &appName,
+		Type:     &appType,
+	}
+
+	app, _, err := appIDClient.RegisterApplicationWithContext(ctx, input)
+
+	if err != nil {
+		return diag.Errorf("Error creating AppID application: %s", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s", tenantID, *app.ClientID))
+
+	return resourceIBMAppIDApplicationRead(ctx, d, meta)
+}
+
+func resourceIBMAppIDApplicationRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	appIDClient, err := meta.(ClientSession).AppIDAPI()
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id := d.Id()
+	idParts := strings.Split(id, "/")
+
+	if len(idParts) < 2 {
+		return diag.Errorf("Incorrect ID %s: ID should be a combination of tenantID/clientID", d.Id())
+	}
+
+	tenantID := idParts[0]
+	clientID := idParts[1]
+
+	app, resp, err := appIDClient.GetApplicationWithContext(ctx, &appid.GetApplicationOptions{
+		TenantID: &tenantID,
+		ClientID: &clientID,
+	})
+
+	if err != nil {
+		if resp != nil && resp.StatusCode == 404 {
+			log.Printf("[WARN] AppID application '%s' is not found", clientID)
+			// we may not want to clear ID here, if user accidentally uses wrong API key it would also get 404
+			// if remote resource is deleted, it should show the error and let the user decide if he wants to taint, replace (recommended)
+			// or remove it from HCL
```

Comment:
I'm not convinced that clearing id on 404 is a good idea, saw other resources doing that, but it seems wrong. I feel like it is better to tell user that something is amiss and let user decide, instead of silently dropping those resources

Reply:
I think this was the approcah followd as per Terraform guidelines ...other providers also has similar handling
https://github.com/hashicorp/terraform-provider-aws/blob/b5600088b1372b07a6852e7419a62af9b40e78a4/aws/resource_aws_api_gateway_api_key.go#L109

Reply:
Hm, wonder what was their reasoning, ok I guess I have no problem putting it back in, just to be consistent